[CLI-457] Removed error when matplotlib plots do not contain data

tssweeney opened the pull request:
Ensures that wandb.Image constructor can accept a pyplot reference in which the figure has multiple axes. Importantly, there is no requirement that any of the axes have plotted data.

* Added a test case to ensure that multiple axes plots will log appropriately
* Removed the check which was causing the false error condition

Fixes: https://github.com/wandb/client/issues/1319
Reference: https://wandb.atlassian.net/browse/CLI-457

adrnswanberg commented on tests/test_data_types.py:
```diff
@@ -241,6 +241,29 @@ def test_matplotlib_image():
+                    if has_data[plot_id]:
+                        ax[plot_id].plot(data)
+            
+            wandb.Image(plt) # this should not error.
```
wandb.Plotly is the other datatype that wraps matplotlib figures. Can you make sure this test covers that too?

tssweeney replied:
Hey @adrnswanberg, great call. I started looking into this and realized that the currently implementation of wandb.Plotly does not accept a matplotlib figure (despite what the docs say). In fact, it only accepts a Plotly figure. Just running some tests now, but the followup commit will have tests and support for such implicit conversion.

tssweeney replied:
Followup from verbal conversation with @adrnswanberg... cc @raubitsj 

Hi Jeff,
Adrian and I just met to discuss this change and wanted to invite you to weigh in. Here are some of the key points:
1. The initial bug was caused because the `util.ensure_matplotlib_figure` method was calling `<pyplot>.gca().has_data()` which fails when the last axes of a plot does not contain data. We made the call to remove this constraint and allow the use to log figures even when data is absent. 
2. As a followup, Adrian pointed out that we should also test the same functionality of the `wandb.Plotly` constructor. Digging in, I found that this constructor does not actually accept pyplot objects (despite the docs). See line 1287 of data_types.py. In response I modified the constructor to make this implicit conversion
3. After speaking with Adrien, we felt it might be questionable to introduce this functionility given that it was not already present, and instead, simply have the test validate that the static `Plotly.make_plot_media` method can accept the pyplot object. This function seems to be the key function used in the recursive serialization of the logged dictionaries. As a result, we would want to remove the documentation stating the the Plotly constructor can accept matplotlib objects.

We would be happy to get your thoughts on this approach, especially as it would be "removing" stated, but not implemented, functionality.

Thanks,
Tim

vanpelt replied:
I think we should make wandb.Plotly accept plotly objects.  It's not a breaking change as it will just add new functionality / fix something that is obviously broken.  I may be confusing plotly for pyplot, but my same logic applies.  Adding this won't break anyones code more than it's already broken.

adrnswanberg replied:
^ I'm ok with this. I hadn't considered closely enough that we do have documentation stating this is supposed to work.

tssweeney replied:
Great! Thanks folks - will move forward with the conversion approach.

adrnswanberg reviewed (approved):
Nice! I like the test. Just one small comment.

tssweeney commented:
Ok, so in order to do this the right way, there was a bit of scope creep. Happy to make this a simpler change if that is desired, but I think the tests add to the robustness. Overall changes:

* Solved the original bug by removing check for data in plots
* Added plotly dev dependency
* Added helper functions (and refactored associated methods) to reduce duplication between wandb.Plotly and wandb.Plotly.make_plot_media:
  - util.matplotlib_to_plotly
  - util.matplotlib_contains_images
* Added additional test cases to test for the following:
  - wandb.Image w/multiple axes optional data
  - wandb.Plotly w/multiple axes optional data
  - wandb.Plotly w/ image
  - wandb.Image w/ image
  - wandb.Plotly.make_plot_media w/ data
  - wandb.Plotly.make_plot_media w/ image
* Added test cases for:
  - util.matplotlib_contains_images w/ and w/o images
  - util.matplotlib_to_plotly 
* Added new tests/utils/dummy_data.py file to help build mock data for tests:
  - matplotlib_multiple_axes_figures
  - matplotlib_with_image
  - matplotlib_without_image

Ultimately there are two reasons for the complexity here: 1) plotly does not support converting matplotlib figures to plotly which contain images (this means we still need to error out if the plotly constructor gets a figure with an image); 2) there is a very similar code path in wandb.Plotly.make_plot_media. I got feedback that there was some non-DRY code in the last commit, so in order to properly refactor, I had to introduce 2 new helper methods and associated tests.

Let me know your thoughts,
Tim

vanpelt reviewed (approved):
This is excellent work, love the tests and refactoring!
